### PR TITLE
Memory machine: Only use the necessary size

### PR DESCRIPTION
**Pull request by georgwiese**
Builds on #1666 

Now, the memory machine is downsized to the smallest possible size:
```
$ MAX_DEGREE_LOG=10 cargo run -r pil test_data/asm/dynamic_vadcop.asm -o output -f --field bn254 --prove-with halo2-mock-composite
...
Running main machine for 1024 rows
[00:00:00 (ETA: 00:00:00)] ░░░░░░░░░░░░░░░░░░░░ 0% - Starting...                         Found loop with period 1 starting at row 100
[00:00:00 (ETA: 00:00:00)] ████████████████████ 100% - 280269 rows/s, 4205k identities/s, 94% progress
Resizing variable length machine 'Secondary machine 0: main_arith (BlockMachine)': 1024 -> 32 (rounded up from 3)
Resizing variable length machine 'Secondary machine 1: main_memory (DoubleSortedWitnesses)': 1024 -> 32 (rounded up from 4)
Witness generation took 0.005317875s
Writing output/commits.bin.
Instantiating a composite backend with 5 machines:
...
== Proving machine: main (size 1024), stage 0
==> Proof stage computed in 15.581958ms
== Proving machine: main__rom (size 16), stage 0
==> Proof stage computed in 800.708µs
== Proving machine: main_arith (size 32), stage 0
==> Proof stage computed in 506.25µs
== Proving machine: main_byte2 (size 65536), stage 0
==> Proof stage computed in 84.450916ms
== Proving machine: main_memory (size 32), stage 0
==> Proof stage computed in 719.375µs
Proof generation took 0.102423415s
Proof size: 88 bytes
Writing output/dynamic_vadcop_proof.bin.
```

### CODE REVIEW

File: pipeline/tests/asm.rs [hunk 1]
```diff
@@ -201,8 +201,8 @@ fn vm_to_block_array() {
 }
 
 #[test]
-fn vm_to_block_different_length() {
-    let f = "asm/vm_to_block_different_length.asm";
+fn dynamic_vadcop() {
+    let f = "asm/dynamic_vadcop.asm";
```
Comment (Schaeff) [inline pipeline/tests/asm.rs:205]: Would this test fail if witgen didn't reduce the degree?

Comment (georgwiese) [reply pipeline/tests/asm.rs:205]: No

File: executor/src/witgen/machines/double_sorted_witness_machine.rs [hunk 2]
```diff
@@ -263,6 +269,21 @@ impl<'a, T: FieldElement> Machine<'a, T> for DoubleSortedWitnesses<'a, T> {
...
+            let new_size = current_size.next_power_of_two() as DegreeType;
+            let new_size = new_size.max(1 << MIN_DEGREE_LOG);
+            log::info!(
+                "Resizing variable length machine '{}': {} -> {} (rounded up from {})",
+                self.name,
+                self.degree,
+                new_size,
+                current_size
+            );
+            self.degree = new_size;
```
[the remark this answers is not in the document]
Comment (georgwiese) [reply executor/src/witgen/machines/double_sorted_witness_machine.rs:285]: Fixed

**Comment by georgwiese** [comment]
Rebased onto `main`

File: pipeline/tests/asm.rs (hunk 1 above)
Comment (Schaeff) [reply pipeline/tests/asm.rs:205]: Is there a way we could test the trace lengths? Could you extract the witness from the pipeline and check its length?

Comment (georgwiese) [reply pipeline/tests/asm.rs:207]: Yeah, good point! Done!

File: pipeline/tests/asm.rs [hunk 3]
```diff
@@ -201,17 +203,26 @@ fn vm_to_block_array() {
...
+
+    let mut pipeline_gl = make_simple_prepared_pipeline(f);
+    let witness = pipeline_gl.compute_witness().unwrap();
+    let witness_by_name = witness
+        .iter()
+        .map(|(k, v)| (k.as_str(), v))
+        .collect::<BTreeMap<_, _>>();
+
+    // Spot-check some witness columns to have the expected length.
+    assert_eq!(witness_by_name["main.X"].len(), 128);
```
Comment (georgwiese) [inline pipeline/tests/asm.rs:217]: Note that that's because we only detect the loop at row 100. But I think that's fine, it's still a small number.

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**